node/network/statement-distribution/src/vstaging/grid.rs

```diff
@@ -0,0 +1,2247 @@
+	/// and that that the manifest includes at least one `Seconded`
+	/// attestation and includes enough attestations for the candidate
+	/// to be backed.
+	///
+	/// This also should only be invoked when we are intended to track
+	/// the knowledge of this peer as determined by the [`SessionTopology`].
+	fn import_received(
+		&mut self,
+		group_size: usize,
+		seconding_limit: usize,
+		candidate_hash: CandidateHash,
+		manifest_summary: ManifestSummary,
+	) -> Result<(), ManifestImportError> {
+		match self.received.entry(candidate_hash) {
+			Entry::Occupied(mut e) => {
+				// occupied entry.
+
+				// filter out clearly conflicting data.
+				{
+					let prev = e.get();
+					if prev.claimed_group_index != manifest_summary.claimed_group_index {
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					if prev.claimed_parent_hash != manifest_summary.claimed_parent_hash {
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					if !manifest_summary
+						.statement_knowledge
+						.seconded_in_group
+						.contains(&prev.statement_knowledge.seconded_in_group)
+					{
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					if !manifest_summary
+						.statement_knowledge
+						.validated_in_group
+						.contains(&prev.statement_knowledge.validated_in_group)
+					{
+						return Err(ManifestImportError::Conflicting)
+					}
+
+					let mut fresh_seconded =
+						manifest_summary.statement_knowledge.seconded_in_group.clone();
+					fresh_seconded |= &prev.statement_knowledge.seconded_in_group;
+
+					let within_limits = updating_ensure_within_seconding_limit(
+						&mut self.seconded_counts,
+						manifest_summary.claimed_group_index,
+						group_size,
+						seconding_limit,
+						&fresh_seconded,
+					);
+
+					if !within_limits {
+						return Err(ManifestImportError::Overflow)
+					}
+				}
+
+				// All checks passed. Overwrite: guaranteed to be
+				// superset.
+				*e.get_mut() = manifest_summary;
+				Ok(())
+			},
+			Entry::Vacant(e) => {
+				let within_limits = updating_ensure_within_seconding_limit(
+					&mut self.seconded_counts,
+					manifest_summary.claimed_group_index,
+					group_size,
+					seconding_limit,
+					&manifest_summary.statement_knowledge.seconded_in_group,
+				);
+
+				if within_limits {
+					e.insert(manifest_summary);
+					Ok(())
+				} else {
+					Err(ManifestImportError::Overflow)
+				}
+			},
+		}
+	}
+}
+
+// updates validator-seconded records but only if the new statements
+// are OK. returns `true` if alright and `false` otherwise.
+//
+// The seconding limit is a per-validator limit. It ensures an upper bound on the total number of
+// candidates entering the system.
+fn updating_ensure_within_seconding_limit(
+	seconded_counts: &mut HashMap<GroupIndex, Vec<usize>>,
+	group_index: GroupIndex,
+	group_size: usize,
+	seconding_limit: usize,
+	new_seconded: &BitSlice<u8, Lsb0>,
+) -> bool {
+	if seconding_limit == 0 {
+		return false
+	}
+
```

Comment:
It's depth + 1 which should never be 0.

Reply:
this is true, but the sanity check doesn't hurt